Week2 [STEP 2] Kobe

devKobe24 opened the pull request:
@Gundy93
 건디, 이번 스텝이 지금 스텝중 가장 어려웠던것 같습니다 ㅠㅠ
 PR 보내드립니다.

 ## 고민했던점
 ### 순서도와 최대한 비슷하게 혹은 똑같게 코드를 만드는 것
 - 순서도가 정말 중요하다고 생각이 들었습니다. 최대한 자세하고 정확한 순서와 제대로된 로직으로 만들지 않으면 코드를 만들 때 더 힘들어진다는 것을 느끼게 되었습니다.

 ---

 ## 해결이 되지 않은 점
 ### 해결이 되지 않은 점은 없었습니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 이번 스텝은 구글 검색을 해가면서 스텝을 이어나갔습니다.
 - 구글 검색을 통한 스텝을 이어나가면 안되는 것일까요?? 괜히 죄책감? 아니면 뭔가 누군가의 로직을 배끼는 듯 한 느낌까지 느껴져서 미안함도 느껴지는데 제가 잘 하고 있는게 맞는지 궁굼합니다.

 ---

Commit:
Week2 [STEP 2] Kobe

 @Gundy93
 건디, 이번 스텝이 지금 스텝중 가장 어려웠던것 같습니다 ㅠㅠ
 PR 보내드립니다.

 ## 고민했던점
 ### 순서도와 최대한 비슷하게 혹은 똑같게 코드를 만드는 것
 - 순서도가 정말 중요하다고 생각이 들었습니다.
 최대한 자세하고 정확한 순서와 제대로된 로직으로 만들지 않으면
 코드를 만들 때 더 힘들어진다는 것을 느끼게 되었습니다.

 ---

 ## 해결이 되지 않은 점
 ### 해결이 되지 않은 점은 없었습니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 이번 스텝은 구글 검색을 해가면서 스텝을 이어나갔습니다.
 - 구글 검색을 통한 스텝을 이어나가면 안되는 것일까요??
 괜히 죄책감? 아니면 뭔가 누군가의 로직을 배끼는 듯 한 느낌까지 느껴져서 미안함도 느껴지는데
 제가 잘 하고 있는게 맞는지 궁굼합니다.

 ---

Gundy93 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -10,3 +10,4 @@ import Foundation
 
 print("Hello, World!")
```
`print("Hello, World!")` 등의 사용하지 않는 코드는 삭제하는 것이 좋겠습니다~

Gundy93 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -10,3 +10,4 @@ import Foundation
 
 print("Hello, World!")
 
+matchLottoNumbers(myLottoNumbers: [1,2,3,4,5,6], winnerLottoNumbers: winnerLottoNumber)
```
`main`에서 선언하지 않은 변수를 사용하면 가독성이 좋지 못한 것 같습니다.

`main`에서 `winnerLottoNumber`를 선언하거나 `matchLottoNumbers` 함수 내부에서 로또번호를 뽑는 함수를 호출해서 비교해도 좋을 것 같습니다.

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+
+import Foundation
+
+var temporaryArr: [Int] = []
```
[Use Terminology Well](https://www.swift.org/documentation/api-design-guidelines/#use-terminology-well)에 따르면 이런 축약형태의 단어를 피하라고 하고, [Promote Clear Usage](https://www.swift.org/documentation/api-design-guidelines/#promote-clear-usage)에 따르면 타입이 아닌 역할에 따라 네이밍을 하라고 합니다.

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+var temporaryArr: [Int] = []
+
+func makeLottoNumbers() -> [Int] {
+	for _ in 0...5 {
+		temporaryArr.append(Int.random(in: 1...45))
+	}
+	
+	var lottoArray = temporaryArr
+	
+	for i in temporaryArr {
+		if lottoArray.contains(i) == false {
+			lottoArray.append(i)
+		}
+	}
+	return lottoArray
```
현재 이 로직으로는 중복 숫자를 막지 못하는 것 같습니다.

<img width="618" alt="스크린샷 2023-03-17 오전 9 31 27" src="https://user-images.githubusercontent.com/106914201/225781733-bb7acdd4-0fdc-4455-9a4a-bd5d0a53b191.png">

`var lottoArray = temporaryArr` 대신 `var lottoArray: [Int] = []`로 바꾸면 중복 숫자는 거를 수 있겠지만 중복시에 숫자가 6개가 되지 못할 것입니다.

순서도대로 만들면 이런 문제가 없을 것입니다.

Gundy93 replied:
`makeLottoNumbers` 함수는 순서도의 흐름과 다른 것 같습니다 🤔

또한 내부 변수 네이밍도 더 고려해야겠습니다.

그리고 함수 외부 변수인 `temporaryArr`가 없다면 이 함수 단품으로는 사용할 수 없겠네요

함수 내부에서 모든 것을 해결하게끔 수정할 수 있을까요?

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+
+let winnerLottoNumber = makeLottoNumbers()
+
+func matchLottoNumbers(myLottoNumbers: [Int], winnerLottoNumbers: [Int]) {
+	var matchLottoNumbers: [Int] = []
```
`matchLottoNumbers`라는 함수명은 함수의 기능을 잘 반영한 네이밍 같습니다.

하지만 변수명 `matchLottoNumbers`는 적절하지 못한 것 같습니다.

`matchedLottoNumbers`로 변경하는 것이 어떨까요?

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+	for number in winnerLottoNumbers {
+		if myLottoNumbers.contains(number) == true {
+			matchLottoNumbers.append(number)
+			print(matchLottoNumbers)
```
현재 `if myLottoNumbers.contains(number) == true`가 참일때마다 `matchLottoNumbers`를 출력하고 있습니다.

테스트 코드였던 것 같은데 커밋 시에는 이러한 테스트 코드를 삭제하고 커밋하는 것이 좋겠습니다!

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+		}
+	}
+	
+	let winnerNumbers = matchLottoNumbers.map { String($0) }.joined(separator: ",")
```
이렇게 할 경우 결과가 다음처럼 나옵니다.

<img width="441" alt="스크린샷 2023-03-17 오전 9 40 52" src="https://user-images.githubusercontent.com/106914201/225782734-8ec74a8d-308e-478c-ac16-6a584cbd5b94.png">

숫자 사이가 공백없이 다닥다닥 붙어있는데요!

쉼표 뒤에 공백이 한 칸 있어야 할 것 같습니다.

Gundy93 commented on CodeStarterCamp_Week2/STEP2.swift:
```diff
@@ -0,0 +1,48 @@
+	
+	let winnerNumbers = matchLottoNumbers.map { String($0) }.joined(separator: ",")
+	
+	if matchLottoNumbers.count >= 1 {
+		print("축하합니다! 겹치는 번호는 \(winnerNumbers) 입니다.")
+	} else if matchLottoNumbers.count == 0 {
+		print("아쉽지만 겹치는 번호가 없습니다.")
+	}
```
`else if`대신 `else`를 써도 같은 결과일 것 같습니다. 이런 경우 `else if`로 추가 조건을 설정하면 이 코드를 읽는 사람이 두 조건을 모두 만족하지 않는 케이스가 있는지 오해할 수 있습니다.

Gundy93 reviewed (commented):
코비님! 기능 구현은 어느정도 잘 된 것 같습니다 👍

> **순서도와 최대한 비슷하게 혹은 똑같게 코드를 만드는 것**
> 순서도가 정말 중요하다고 생각이 들었습니다. 최대한 자세하고 정확한 순서와 제대로된 로직으로 만들지 않으면 코드를 만들 때 더 힘들어진다는 것을 느끼게 되었습니다.

코드와 무관하게 자연스러운 흐름을 만들어내는 능력을 기르면 그것을 코드로 옮기는 것은 쉬우니까요! 연습을 더 해봅시다

> **이번 스텝은 구글 검색을 해가면서 스텝을 이어나갔습니다.**
> 구글 검색을 통한 스텝을 이어나가면 안되는 것일까요?? 괜히 죄책감? 아니면 뭔가 누군가의 로직을 배끼는 듯 한 느낌까지 느껴져서 미안함도 느껴지는데 제가 잘 하고 있는게 맞는지 궁굼합니다.

원래 한 사람이 Swift의 모든 기능을 알 수 없기 때문에 원하는 특정 기능이 있는지 검색하는 것은 자연스러운 일입니다!

중요한 것은 '코드의 어떤 부분이 어떤 기능을 하기때문에 이러한 흐름으로 코드가 진행되고, 나는 여기서 이런 부분을 참고하면 좋겠다'와 같은 판단을 적절히 하는 것입니다.

모르는 부분은 검색하는 것이 자연스러운 흐름입니다.

이번 수정도 화이팅입니다!

Commit:
Week2 [STEP 2] Kobe
 @Gundy93
 건디, 벌써 마지막 주 입니다 ㅠㅠ
 더 힘내서 피드백 받은 내용을 중심으로 수정하여 다시 PR 보냅니다.

 ## 고민했던 점
 ### 순서도와 비슷하게 혹은 최대한 똑같게 코드를 만드는 것
 - 이전 PR 과 동일하게 순서도와 최대한 똑같게 코드를 만들려고 노력했습니다.
 어떻게 하면 순서도와 비슷한 코드가 혹은 똑같은 코드가 나올까 생각하며 코드를 만들었습니다.

 - 고정 관념이였던 꼭 배열만 써야 한다는 것을 깨고 세트를 활용하고 for문 말고 while 문을 써봤습니다.
 제 생각속에 갇혀서 꼭 배열과 for문 if 문만 써야 한다는 생각에서 탈피해 while 문과 set을 활용하여 순서도처럼 코드를 만들어봤습니다.

 ---

 ## 해결 되지 않은 점
 ### 해결이 되지 않은 점은 없었습니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 이번 스탭 역시 구글 서칭을 이용했습니다.
 - 지난번 피드백에서 구글 서칭은 자연스러운 일이라고 말씀해주셔서 구글 서칭을 활용하여
 많은 지식을 얻었고 그로 인해 코드를 이번 스텝을 해결하게 되었습니다.
 구글 서칭을 하고 난 뒤에 건디는 따로 정리를 하시는 편인가요??
 그것이 궁굼합니다 :)

Commit:
Week2 [STEP 2] Kobe

 @Gundy93
 건디, 피드백 받은대로 수정했습니다 :)

 ## 고민했던 점
 ### 네이밍에 대해서 고민을 하게 되었습니다.
 - 네이밍에 대해서 조금 더 신중하게 생각하고 고민했습니다.

 ---

 ## 해결 되지 않은 점
 ### 해결이 됮 않은 점은 없었습니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 얼마남지 않은 기간 더 열심히 하겠습니다.
 - 조언보다는 지금 더 열심히 하겠다는 의지를 보여드리고 싶었습니다 :)
